man/man1/host.1
.TH host 1 "January 28, 2021" "Version 1.2.1"
.SH NAME
host - a simple, static, web server for easily hosting files over LAN

.SH SYNOPSIS
host --help

host [-h] [-q] FILE

host [--hidden] [--quiet] [--output OUTPUT_FILE] [--port PORT] FILE

host -i [-h] [-q]

host --stdin [--hidden] [--quiet] [--output OUTPUT_FILE] [--port PORT]

.SH DESCRIPTION

host is a simple, static, web server for easily hosting files over LAN. host
makes no security guarantees, so it should only be used on a LAN that you trust
(no port forwarding and probably no coffee shops).

host runs in three different modes: file, stdin, and directory.

.PP
In file mode, the server responds to GET requests with the contents of the
file, regardless of the route sent. In file mode, the program reads the
contents of the file every time a GET request is sent.

Stdin mode acts similarly to file mode, just that host reads from stdin instead
of a file and stdin is only read once. Because of this read-once property of
host, it could be used as a method to host a file without changes on disk
affecting changes in what is being hosted.

.PP
Directory mode is very different. In directory mode, the server acts like a
real server, allowing the clients to traverse the directory tree, where the
root of the directory tree is the specified directory (see FILE above). host
has an html template, known as directory.html, embeded inside. This html
template is invoked every time a user sends a request to a directory that does
not contain an index.html file, when the user adds the ls key to their query
(localhost:PORT/foobar?ls), or when the user requests a file that does not 
exist.

.PP
These modes are selected at runtime based on the command line arguments passed
into host. If the -i or --stdin option is selected, host will execute in stdin
mode. If a file is passed into the FILE argument (see SYNOPSIS), host will run
in file mode. Likewise, if a directory is passed into the FILE argument, host
will run in directory mode.

.PP
The way host handles mimetypes depends on the mode host is in. If host is in
directory or file mode, host will respond with the mimetype appropriate for
the file, based on the file extension, with the default being text/plain.
If host is started in stdin mode, host will respond with the text/plain
mimetype.

Using the --mime option, and if host is in file or stdin mode, host will
start with the mimetype described. Note: the --mime option expects file
extentions, so use html instead of text/html, txt instead of
text/plain, etc

.SH OPTIONS

.TP
.B --help

Displays help text and exits

.TP
.B -h, --hidden

Shows hidden files. This option is only useful in directory mode and will do
nothing if it is outside of directory mode.

.TP
.B -i, --stdin

Puts host into stdin mode and will serve whatever it finds from stdin.

.TP
.B --mime [mimetype]

Sets the mimetype that will be sent by host. This is useful for stdin mode, as
the default mimetype is text/plain. In file mode, the mimetype is determined
via the file extension. This option overrides that. Mimetypes should be
inputted via their file extensions and not the `type/standard` format.
(so, html over text/html, txt over text/plain, etc.)

.TP
.B --output [FILE]

Instead of printing out request logs, the request logs will be silently sent
to FILE.

.TP
.B --port [PORT]

Set the port host will be host on. Set to 8080 by default.

.TP
.B -q, --quiet

Prevents host from printing request logs.

.SH AUTHOR

Please send all bug reports to the GitHub issue tracker:
.B https://github.com/RainbowAsteroids/host/issues

host was written by Rainbow Asteroids (rainbowasteroids@protonmail.com) in Nim.


src/host.nim
import asynchttpserver, asyncdispatch
import strutils, os, re, algorithm, tables
import mimetypes, strformat, parseopt, parseutils
import net

# Embed some important resources
const directory_html = slurp"../assets/directory.html"
const file_html = slurp"../assets/file.html"
const help_text = """Usage: host [options] [file]

Options:
  --help              : Show this help text
  -h, --hidden        : Shows hidden files. Only useful in directory mode.
  -i, --stdin         : Host from stdin. Ignores file parameter.
  --mime [mimetype]   : Set the mimetype of the file/stdin. Ignored in directory mode.
  --output [file]     : Output request logs to a file.
  --port [port]       : Host on a specific port (default 8080)
  -q, --quiet         : Hide request logs."""

type
  InputType = enum
    itFile,
    itDir,
    itStdIn

  Input = object
    case inputType: InputType
    of itFile, itDir: path: string
    of itStdIn: str: string

type
  RequestLogType = enum
    rltEcho, rltNone, rltFile

  RequestLogState = object
    case rlType: RequestLogType
    of rltFile: file: File
    else: discard

var showHidden = false
var requestLogState = RequestLogState(rlType: rltEcho)

proc combineDir(sub, tail: string): string =
  let tail = if tail[0] == '/': tail else: '/' & tail
  if sub[^1] == '/': sub[0..^2] & tail
  else: sub & tail

proc logRequest(req: Request) =
  if requestLogState.rlType != rltNone:
    let log = fmt"req: {req.url.path} ; by: {req.hostname}"
    if requestLogState.rlType == rltFile:
      requestLogState.file.write(log & '\n')
    else:
      echo log

proc generateHeader(mime: string): HttpHeaders =
      [("content-type", fmt"{mime}; charset=UTF-8")].newHttpHeaders

proc generateRequestHandlerFromString(data, mime: string): proc =
  let header = generateHeader(mime)
  proc requestHandler(req: Request) {.async.} =
    req.logRequest()
    await req.respond(Http200, data, header)
  requestHandler

proc generateRequestHandlerFromFile(fileName, mime: string): proc =
  let header = generateHeader(mime)
  proc requestHandler(req: Request) {.async.} =
    req.logRequest()
    await req.respond(Http200, readFile(fileName), header)
  requestHandler

proc generateRequestHandlerFromDirectory(dir: string): proc =
  proc requestHandler(req: Request) {.async.} =
    # 0. Check that the user isn't requesting outside of our dir
    # 1. Check if we should invoke directory.html
    # 2. Check if requested file is a directory
    # 2a. No, send file
    # 3. If it is a directory, look for index.html
    # 3a. If found, send index.html
    # 3b. If not found, send directory.html
    # 4. If nothing is found, send a 404

    type Warning = enum
      wNoIndex = "no index.html found"
      wRequested = "directory.html requested"
      wNotFound = "404 - file not found"

    type File = object
      name, html: string
      isDir: bool

    proc newFile(name, html: string, isDir: bool): File =
      result.name = name
      result.html = html
      result.isDir = isDir

    var workingPath = combineDir(dir, req.url.path)
    workingPath = workingPath.replace(re"%20", " ")

    req.logRequest()

    proc DirectoryHtmlRequested(query: seq[string]): bool =
      for q in query:
        if q.split(re"=")[0] == "ls":
          return true

      return false

    proc generateDirectoryDotHtml(warn: Warning, workingPath, rawReqPath: string): string =
      result = directory_html.replace("{warn}", $warn)
      result = result.replace("{path}", rawReqPath.replace(re"%20", " "))
      var files: seq[File]
      var htmlToInsert: string

      for kind, path in walkDir(workingPath, true):
        if not showHidden and path[0] == '.':
          continue # Let's not show showHidden files if we aren't told to

        let truePath = combineDir(workingPath, path) # Path we need to use
        let reqPath = combineDir(rawReqPath, path) # Path the client uses
        let isDir = dirExists(truePath) # isDir field for file obj

        var html = file_html.replace("{file_path}", reqPath)
        html = html.replace("{file_name}", if isDir: path & '/' else: path)

        files.add(newFile(path, html, isDir))

      files.sort do (x, y: File) -> int:
        if x.isDir and not y.isDir:
          -1
        elif y.isDir and not x.isDir:
          1
        else:
          cmp(x.name.toLowerAscii, y.name.toLowerAscii)

      for f in files:
        htmlToInsert = htmlToInsert & f.html

      result = result.replace("{files}", htmlToInsert)

    if req.url.path.contains(".."):
      await req.respond(Http400, "Error 400: why are you using `..`?",
                        headers = generateHeader("text/plain"))

    elif not (req.url.query == "") and
         DirectoryHtmlRequested(req.url.query.split("&")):
      if fileExists(workingPath):
        workingPath = parentDir(workingPath)
      await req.respond(Http200, generateDirectoryDotHtml(wRequested, workingPath, req.url.path),
                        headers = generateHeader("text/html"))

    elif dirExists(workingPath):
      let indexPath = combineDir(workingPath, "/index.html")
      if fileExists(indexPath):

        await req.respond(Http200, readFile(indexPath),
                          headers = generateHeader("text/html"))
      else:
        await req.respond(Http200, generateDirectoryDotHtml(wNoIndex, workingPath, req.url.path),
                          headers = generateHeader("text/html"))

    elif fileExists(workingPath):
      let m = newMimeTypes()
      await req.respond(Http200, readFile(workingPath),
                        headers = generateHeader(m.getMimeType(workingPath.splitFile.ext)))
    else:
      await req.respond(Http404, generateDirectoryDotHtml(wNotFound, 
                                                  workingPath.splitFile.dir, 
                                                  req.url.path.splitFile.dir),
                        headers=generateHeader("text/html"))

  requestHandler

proc main(input: Input, port: Natural, mime: string) =
  var mime = mime
  var server = newAsyncHttpServer()
  let mimedb = newMimetypes()

  echo "Starting server with IP ", $getPrimaryIPAddr(), " and port ", port
  case input.inputType:
    of itFile:
      if mime != "":
        mime = mimedb.getMimeType(mime)
      else:
        mime = mimedb.getMimeType(input.path.splitFile.ext)
      waitFor server.serve(Port(port),
                           generateRequestHandlerFromFile(input.path, mime))
    of itDir:
      waitFor server.serve(Port(port),
                           generateRequestHandlerFromDirectory(input.path))
    of itStdIn:
      mime = mimedb.getMimeType(mime)
      waitFor server.serve(Port(port),
                           generateRequestHandlerFromString(input.str, mime))

when isMainModule:
  var port = 8080
  var stdinMode = false
  var filename: string
  var mime = ""

  var p = initOptParser(shortNoVal = {'i', 'h', 'q'},
                        longNoVal = @["stdin", "hidden", "quiet"])
  for kind, key, val in p.getOpt():
    case kind
      of cmdArgument:
        filename = key
      else:
        case key
          of "stdin", "i":
            stdinMode = true
          of "hidden", "h":
            showHidden = true
          of "quiet", "q":
            requestLogState = RequestLogState(rlType: rltNone)
          of "output":
            requestLogState = RequestLogState(rlType: rltFile,
                                              file: open(val, fmWrite))
          of "port":
            discard parseInt(val, port)
          of "mime":
            mime = val
          else: # or `of "help"`
            if key != "help": echo fmt"I didn't understand option `{key}`!\n"
            echo help_text
            quit()

  if stdinMode:
    main(Input(inputType: itStdIn, str: stdin.readAll()), port.Natural, mime)
  else:
    if filename == "":
      echo "No filename passed in!\n"
      echo help_text
    elif fileExists(filename):
      main(Input(inputType: itFile, path: filename), port.Natural, mime)
    elif dirExists(filename):
       main(Input(inputType: itDir, path: filename), port.Natural, mime)
    else:
      echo fmt"File {filename} not found!"

